Add Codeblock Syntax Highlighting using Pygments

Rohitth007 opened the pull request:
This PR implements syntax highlighting for markdown code blocks by creating a padded "box" which would
visually separate as code. Currently, only looks good in dark theme.

 **Commit 1:** themes: codehilite: Adding Pygments style to the palette.
* This commit uses Pygments internal dictionaries and Pygments
Material Style and adds it into the palette after the Screen is
initialized. This is done inside `add_codehilite_style` and called
in `Controller.__init__`.

**Commit 2:** boxes: codehilite: Implement syntax highlighting functionality.
* Basic syntax highlighting functionality was implemented in
soup2markup by extracting the `css_style` of each `span` tag in the
soup and using it as urwid style attributes. The remaining
NavigableStrings are given a style attribute of `w` which is the
whitespace css class in Pygments.

**Commit 3:** boxes: codehilite: Create a visual "box" by padding whitespaces.
* This commit moves the previous code into `parse_codeblock`, a static
method which creates a visual "box" by:
   * Finding the longest line using `code_soup.text`.
   * Appending `span` text into markup
   * Adding `code_indent` part of the NavigableString to markup.
   * Adding `right_padding` to each line.
   * Making sure multiple empty lines are padded using `padded_line`.
   * Adding the remaining NavigableStrings not contributing to line
breaks into markup.
* The NavigableString is added in parts and not all at once because
though it would produce the same output, it would later help while
implementing `hl_lines` preventing the case where half the previous
line is highlighted along with the line expected.

**Commit 4:** boxes: codehilite: Implement padding for Plain Codeblocks.
* Plain Codeblock have to be treated differently as they don't come in
the usual multiple `span`'s in a `pre` tag package. The whole code
is a NavigableString which does not start with `\n`.
* Hence, if this String has multiple lines it is split and padded
accordingly.

**Commit 5:** boxes: codehilite: Restrict padding to avoid wrapping.
* This commit sets a limit to how much padding is to be added. Many a
times most of the code has smaller lines and there is one line which
is very long which causes all the lines to wrap and this makes the
"box" look very messy. To avoid this a `MAX_NO_WRAP_WIDTH` constant
of 80 is used which is a reasonable limit for in a full screen
no autohide mode.

neiljp commented:
@Rohitth007 This doesn't run for me - you don't seem to have added the dependency information in a commit? I could guess, but please add it as a separate commit for now based on what you have installed locally.

Rohitth007 commented:
@neiljp Strange, can I know what the error was because I haven't installed anything separately. `Pygments` installs as a dependency to `pudb`

Rohitth007 commented:
I have added Pygments as a dependency into `setup.py` and also fixed the edge case for old messages without `code` tag.

Rohitth007 commented:
@zulipbot add "feedback wanted"

neiljp commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -1419,6 +1414,44 @@ def indent_quoted_content(soup: Any, padding_char: str) -> int:
+        max_line_len = len(max(code_text.split("\n"), key=lambda x: len(x)))
+
+        line_len = 0
+        for code_element in code_soup.contents:
```
The logic here looks very complex, and I can't tell if it's doing the correct thing. Can you loop over lines instead?

Rohitth007 replied:
No that is not possible, we have to parse element wise. Sometimes that element can contain only one character only. I think  if we split lines using code_text we would loose out on tag information(css class, element). I can write a docstring explaining the `codeblock` format?

neiljp reviewed (commented):
@Rohitth007 This works much better for me now :+1:

I can still observe some bugs, which might be clearer where they come from once you add some tests and if you can perhaps simplify the padding algorithm: (these are more/only visible somewhere after the latter 3 commits)
* See first line of last message by 'Mehdi Hamidi' in #**test here>codeblock** (user list is disrupted)
* See message by 'Vishwesh JainKuniya' in #**test here>code** on July 31 2017 (bottom line content is absent)

I think the first three commits are close to merging as they are, pending at least the following:
* given we're modifying the behavior for code, I'd like to see the soup2markup test updated for this
* using the pygments: namespace to avoid accidental style name conflicts
* improving the commit title/text to match the styles we have (eg. see the requirements lines we use for dependencies)
* clarifying what the 'old' vs 'new' messages categories are
* I'm not sure what we can do about inline code styles, but we likely want them to match in some way at least

Rohitth007 commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -1118,7 +1118,19 @@ def soup2markup(
-                markup.append(("msg_code", tag_text))
+                pre_tag = element.contents[0]
+                code_soup = pre_tag.contents[1]
+                # NOTE: Old messages don't have the
+                # additional `code` tag.
+                if code_soup.name != "code":
+                    code_soup = pre_tag
```
No this is only from observation

neiljp replied:
We should definitely clarify when this occurs, or used to occur. Do you have examples of old and new messages exhibiting this? Given we can look at the source of them, is this literally for messages before some date, or is it a different style of markdown?

Rohitth007 replied:
Examples:
* https://chat.zulip.org/#narrow/stream/7-test-here/topic/code/near/130791
* https://chat.zulip.org/#narrow/stream/7-test-here/topic/html.20elements.20in.20code.20blocks/near/730016
Possibly before April 2019 atleast
Example of same message before and after:
```html
<pre><span></span>&amp;lt;span&amp;gt;test&amp;lt;/span&amp;gt;
</pre>

<pre><span></span><code>&amp;lt;span&amp;gt;test&amp;lt;/span&amp;gt;
</code></pre>
```

neiljp replied:
It would be good to get a point of reference for this - maybe a czo discussion exists? The server changelog or commits may include this, but it may be hard to find, so start a new topic on eg #backend if you can't find a reference.

Rohitth007 commented:
PR updated, tests added and ready for another review.
I have simplified the code a bit and added comments and doc-strings to help understand the HTML structure.
Minor addition: Used Pygments style for inline code

neiljp reviewed (commented):
@Rohitth007 I'm focusing on the first three first, since they should be able to be merged together first, possibly with the new last commit if we reorder them?

The tests are a good improvement, and also document what tags are doing (see minor inline comment).

My biggest concern for the first 3 (and maybe last one?) is the style behavior, including at different depths - both ensuring they're consistent with what we have now (monochrome; see inline comment) and have good contrast with the surrounding text (like now). Things like python, js or bash tagged code look colored great, but with no language tag it just looks like regular text, not a code block - and so it doesn't stand out from the surrounding text that well. I understand we expect to have theme-specific pygments styles, but we want it to keep working well close to how it is for now.

I'd also like to have a reference for that change in syntax (`<code>`) in a comment - various parts of the community may know the origin of that if it's not obvious in discussions/logs.

In the next iteration, improving the commit title/text to match the styles we have will make it easier and quicker to merge commits (I think I mentioned that above)

Rohitth007 commented on tests/ui/test_ui_tools.py:
```diff
@@ -1397,9 +1397,96 @@ def test_private_message_to_self(self, mocker):
-                id="codehilite",
+                '<div class="codehilite" data-code-language="python">'
+                "<pre><span></span>"
+                "<code><span>def</span> <span>func</span><span>():</span>\n"
+                '    <span class="pg">print</span><span>()</span>\n'
+                "\n"
+                "<span>class</span> <span>New</span><span>:</span>\n"
+                '    <span>name</span> <span>=</span> <span>"name"</span>\n'
```
I wanted to test multiple empty lines and spaces, hence the longer tests. I don't think it matters what "text/syntax" is used for testing here as pygments is not directly involved.

Rohitth007 commented on zulipterminal/config/themes.py:
```diff
@@ -586,3 +589,20 @@ def theme_with_monochrome_added(theme: ThemeSpec) -> ThemeSpec:
+
+def add_codehilite_style(screen: Any) -> None:
+    screen.register_palette_entry(
+        "pygments:w", "", "", None, DEF["white"], DEF["black"]
```
I'm not considering the 2nd commit ready yet as I had randomly chosen colors and used hacky methods to get
pygments to work (startswith("#")) hence some lines in the code look different.

Rohitth007 commented on tests/ui/test_ui_tools.py:
```diff
@@ -1397,9 +1397,96 @@ def test_private_message_to_self(self, mocker):
+                [
+                    ("pygments:w", "This is a\n    Plain\n\n    Codeblock\n"),
+                ],
+                id="codehilite-plain-codeblock",
```
I have used `plain-text-codeblock` instead

neiljp commented:
@Rohitth007 I'm thinking of close this, since some of these commits were merged already. I'd suggest you open a new PR with the remaining work after fetching and rebasing, so we can focus on those.

That said, I know it wasn't clear what to do regarding long lines and we ended up with various options based on spaces, which wraps badly. Do you think a widget-based approach might work better?

neiljp commented:
@Rohitth007 Thanks for all the work on the syntax highlighting! :+1:

I've just merged #1346 with a version of the last commit from here, which sets the inline style to match codeblocks, though I adjusted the existing theme style name rather than removing it, since we still use it elsewhere.

We discussed flattening the edges of the code blocks in a separate PR, but I don't believe anything came of that, so I've opened #1347 for that remaining element. That points back here for reference, so I'm going to close this now, since I recall that was the remaining element we hadn't merged.
